compiler/src/dotty/tools/dotc/staging/CrossStageSafety.scala

```diff
@@ -74,21 +82,17 @@ class CrossStageSafety extends TreeMapWithStages {
       case tree @ QuotedTypeOf(body) =>
         if (ctx.property(InAnnotation).isDefined)
           report.error("Cannot have a quote in an annotation", tree.srcPos)
-        body.tpe match
-          case DirectTypeOf(termRef) =>
-            // Optimization: `quoted.Type.of[x.Underlying](quotes)`  -->  `x`
-            ref(termRef).withSpan(tree.span)
+
+        val (tags, body1) = transformQuoteBody(body, tree.span)
+        tags match
+          case tag :: Nil if body1.isType && body1.tpe =:= tag.tpe.select(tpnme.Underlying) =>
+            tag // Optimization: `quoted.Type.of[x.Underlying](quotes)`  -->  `x`
           case _ =>
-            transformQuoteBody(body, tree.span) match
-              case DirectTypeOf.Healed(termRef) =>
-                // Optimization: `quoted.Type.of[@SplicedType type T = x.Underlying; T](quotes)`  -->  `x`
-                ref(termRef).withSpan(tree.span)
-              case transformedBody =>
-                val quotes = transform(tree.args.head)
-                // `quoted.Type.of[<body>](quotes)`  --> `quoted.Type.of[<body2>](quotes)`
-                val TypeApply(fun, _) = tree.fun: @unchecked
-                if level != 0 then cpy.Apply(tree)(cpy.TypeApply(tree.fun)(fun, transformedBody :: Nil), quotes :: Nil)
-                else tpd.Quote(transformedBody).select(nme.apply).appliedTo(quotes).withSpan(tree.span)
+            // `quoted.Type.of[<body>](<quotes>)` --> `'[<body1>].apply(<quotes>)`
+            val quotes = transform(tree.args.head)
+            val TypeApply(fun, _) = tree.fun: @unchecked
+            if level != 0 then cpy.Apply(tree)(cpy.TypeApply(tree.fun)(fun, body1 :: Nil), quotes :: Nil)
```

Comment:
I was confused when I first read this code, because `tags` isn't used in this branch, I had to go read transformQuoteBody to realize that when level != 0 then tags is always Nil. Maybe, adding a comment mentioning this would be enough.

Reply:
I inlined `transformQuoteBody` and refactored the code to only make this transformation at `level == 0`. Other levels are transformed using `super.transform`.